### PR TITLE
Feature gsgmf 7 provide tracking

### PR DESCRIPTION
**Pull request by apastuhov**
https://jira.camptocamp.com/browse/GSGMF-7

This PR is only about the autorotating feature.

### CODE REVIEW

File: contribs/gmf/src/controllers/abstract.js [hunk 1]
```diff
@@ -651,4 +655,149 @@ gmf.AbstractController.prototype.search_ = function(query, overlay) {
...
+      hdg = hdg + (Math.PI / 2);
+      break;
+
+    case 270:
+      hdg = hdg - (Math.PI / 2);
+      break;
+
+    default:
+      hdg = hdg;
+  }
```
Comment (sbrunner) [inline contribs/gmf/src/controllers/abstract.js]: All the switch =>  `hdg = hdg + orientation * Math.PI / 180` not ?

File: contribs/gmf/src/controllers/abstract.js [hunk 2]
```diff
@@ -651,4 +655,149 @@ gmf.AbstractController.prototype.search_ = function(query, overlay) {
...
+  let heading = headingFromDevices(this.deviceOrientation);
+  if (angular.isDefined(heading)) {
+    heading = -heading;
+    let currRotation = this.map.getView().getRotation();
+    const diff = heading - currRotation;
+
+    if (diff > Math.PI) {
+      heading -= 2 * Math.PI;
+    } else if (diff < -Math.PI) {
+      currRotation -= 2 * Math.PI;
```
Comment (sbrunner) [inline contribs/gmf/src/controllers/abstract.js]: not used ..., => `heading`?

File: contribs/gmf/src/controllers/abstract.js [hunk 3]
```diff
@@ -651,4 +655,149 @@ gmf.AbstractController.prototype.search_ = function(query, overlay) {
...
+
+  this.deviceOrientation.on(['change'], (event) => {
+    const heading = headingFromDevices(this.deviceOrientation);
+    if (!angular.isDefined(heading)) {
+      console.error('Heading is undefined');
+      return;
+    }
+
+    if (heading < currHeading - 0.05 ||
+        currHeading + 0.05 < heading) {
```
Comment (sbrunner) [inline contribs/gmf/src/controllers/abstract.js]: => `if (abs(heading - currHeading) > 0.05) {`...

File: contribs/gmf/src/controllers/abstract.js [hunk 4]
```diff
@@ -651,4 +655,149 @@ gmf.AbstractController.prototype.search_ = function(query, overlay) {
...
+
+  // Geolocation control events
+  // geolocation.on('change:position', function(evt) {
+  //   bt.removeClass('ga-btn-disabled');
+  //   locate();
+
+  //   updatePositionFeature();
+  //   updateAccuracyFeature();
+  //   updateHeadingFeature();
+  // });
```
Comment (sbrunner) [inline contribs/gmf/src/controllers/abstract.js]: Debugging code to remove...

File: contribs/gmf/src/controllers/abstract.js [hunk 5]
```diff
@@ -651,4 +655,149 @@ gmf.AbstractController.prototype.search_ = function(query, overlay) {
...
+
+// Orientation control events
+gmf.AbstractController.prototype.autorotateListener = function() {
+  this.deviceOrientation = new ol.DeviceOrientation();
+
+  console.log(1, this.deviceOrientation);
+
+  let currHeading = 0;
+  const headngUpdateWhenMapRotate = throttle(this.headingUpdate, 300, this);
+  // let headngUpdateWhenIconRotate = gaThrottle.throttle(headingUpdate, 50);
```
Comment (sbrunner) [inline contribs/gmf/src/controllers/abstract.js]: Code to remove

File: contribs/gmf/src/controllers/abstract.js [hunk 6]
```diff
@@ -651,4 +655,149 @@ gmf.AbstractController.prototype.search_ = function(query, overlay) {
...
+}
+
+// Get heading depending on devices
+function headingFromDevices(deviceOrientation) {
+  let hdg = deviceOrientation.getHeading();
+  let orientation = window.orientation;
+  if (hdg === undefined) {
+    return undefined;
+  }
+  if (angular.isDefined(hdg) && !isIOS()) {
```
Comment (sbrunner) [inline contribs/gmf/src/controllers/abstract.js]: `angular.isDefined(hdg)` is always true, not?

**Comment by fredj** [comment]
@apastuhov please change all `angular.isDefined(val)` with `val !== undefined`

File: contribs/gmf/src/controllers/abstract.js (hunk 6 above)
Comment (apastuhov) [inline contribs/gmf/src/controllers/abstract.js]: +

File: contribs/gmf/src/controllers/abstract.js (hunk 1 above)
Comment (apastuhov) [inline contribs/gmf/src/controllers/abstract.js]: no, look at the case of orientation 270

**Comment by apastuhov** [comment]
@sbrunner and @fredj thanks for review, PR was updated according to your requests

File: contribs/gmf/src/controllers/abstract.js (hunk 1 above)
Comment (sbrunner) [inline contribs/gmf/src/controllers/abstract.js]: Then:  `hdg = hdg + ((orientation + 179) % 360 - 179)  * Math.PI / 180`
I think that it should also working with (depends of the rest of the code...): 
`hdg = hdg + orientation % 360  * Math.PI / 180`

Comment (apastuhov) [inline contribs/gmf/src/controllers/abstract.js]: It may work, but it is not obvious at all...

**Comment by apastuhov** [comment]
@sbrunner , I moved autorotation inside of geolocation directive so now feature is working as expected with rotation and tacking. Can I ask you to check it on your devices?

Example of mobile-geolocation with tracking and autorotation:

https://camptocamp.github.io/ngeo/feature-GSGMF-7-provide-tracking/examples/mobilegeolocation.html

I can not check mobile app, because I am not in Switzerland and geolocation does not work for me:

https://camptocamp.github.io/ngeo/feature-GSGMF-7-provide-tracking/examples/contribs/gmf/apps/mobile/

File: src/directives/mobilegeolocation.js [hunk 7]
```diff
@@ -216,11 +230,11 @@ ngeo.MobileGeolocationController = function($scope, $element,
...
-  ol.events.listen(
-    view,
-    ol.Object.getChangeEventType(ol.ViewProperty.ROTATION),
-    this.handleViewChange_,
-    this);
+  // ol.events.listen(
+  //   view,
+  //   ol.Object.getChangeEventType(ol.ViewProperty.ROTATION),
+  //   this.handleViewChange_,
+  //   this);
```
Comment (sbrunner) [inline src/directives/mobilegeolocation.js]: Pleas remove this commented code ...

File: contribs/gmf/src/controllers/abstract.js (hunk 1 above)
Comment (sbrunner) [inline contribs/gmf/src/controllers/abstract.js]: one switch with 5 entry is clearly worst!
and you can split it to be more readable like:
```js
// Normalize to be between -90 and 180 
orientation =  ((orientation + 179) % 360 - 179);
// Add to hdg in radian
hdg += orientation * Math.PI / 180
```

**Comment by apastuhov** [comment]
@sbrunner updated

File: src/directives/mobilegeolocation.js [hunk 8]
```diff
@@ -312,5 +320,120 @@ ngeo.MobileGeolocationController.prototype.handleViewChange_ = function(event) {
...
+      easing: ol.easing.linear
+    });
+  }
+};
+
+// Orientation control events
+ngeo.MobileGeolocationController.prototype.autorotateListener = function() {
+  this.deviceOrientation = new ol.DeviceOrientation();
+
+  console.log(1, this.deviceOrientation);
```
Comment (fredj) [inline src/directives/mobilegeolocation.js]: to remove: it breaks the travis checks

**Pull request merged**